lib/AST/DiagnosticEngine.cpp

```diff
@@ -498,56 +563,31 @@ static void formatDiagnosticArgument(StringRef Modifier,
     Out << FormatOpts.ClosingQuotationMark;
     break;
 
-  case DiagnosticArgumentKind::Type: {
+  case DiagnosticArgumentKind::Type:
     assert(Modifier.empty() && "Improper modifier for Type argument");
-    
-    // Strip extraneous parentheses; they add no value.
-    auto type = Arg.getAsType()->getWithoutParens();
-
-    // If a type has an unresolved type, print it with syntax sugar removed for
-    // clarity. For example, print `Array<_>` instead of `[_]`.
-    if (type->hasUnresolvedType()) {
-      type = type->getWithoutSyntaxSugar();
-    }
-
-    bool isAmbiguous = typeSpellingIsAmbiguous(type, Args);
-
-    if (isAmbiguous && isa<OpaqueTypeArchetypeType>(type.getPointer())) {
-      auto opaqueTypeDecl = type->castTo<OpaqueTypeArchetypeType>()->getDecl();
-
-      llvm::SmallString<256> NamingDeclText;
-      llvm::raw_svector_ostream OutNaming(NamingDeclText);
-      auto namingDecl = opaqueTypeDecl->getNamingDecl();
-      if (namingDecl->getDeclContext()->isTypeContext()) {
-        auto selfTy = namingDecl->getDeclContext()->getSelfInterfaceType();
-        selfTy->print(OutNaming);
-        OutNaming << '.';
-      }
-      namingDecl->getFullName().printPretty(OutNaming);
-
-      auto descriptiveKind = opaqueTypeDecl->getDescriptiveKind();
-
-      Out << llvm::format(FormatOpts.OpaqueResultFormatString.c_str(),
-                          type->getString().c_str(),
-                          Decl::getDescriptiveKindName(descriptiveKind).data(),
-                          NamingDeclText.c_str());
-
-    } else {
-      auto printOptions = PrintOptions();
-      printOptions.FullyQualifiedTypes = isAmbiguous;
-      std::string typeName = type->getString(printOptions);
-
-      if (shouldShowAKA(type, typeName)) {
-        llvm::SmallString<256> AkaText;
-        llvm::raw_svector_ostream OutAka(AkaText);
-        OutAka << type->getCanonicalType();
-        Out << llvm::format(FormatOpts.AKAFormatString.c_str(),
-                            typeName.c_str(), AkaText.c_str());
-      } else {
-        Out << FormatOpts.OpeningQuotationMark << typeName
-            << FormatOpts.ClosingQuotationMark;
+    formatTypeArgument(
+        Arg.getAsType(),
+        [Args](Type Ty) { return typeSpellingIsAmbiguous(Ty, Args); },
+        FormatOpts, Out);
+    break;
+  case DiagnosticArgumentKind::TypeDescription: {
+    assert(Modifier.empty() &&
+           "Improper modifier for TypeDescription argument");
+    auto description = Arg.getAsTypeDescription();
+    auto isAmbiguous = [description](Type Ty) {
+      if (auto nominal = Ty->getAnyNominal()) {
+        UnqualifiedLookup lookup(nominal->getFullName(), description.DC,
```

Comment:
I noticed there have been a lot of changes to name lookup lately, so I'm not entirely sure this is the best way to do the ambiguity check.

Reply:
I think we really shouldn't do lookup during diagnostic formatting. That very much weirds me out and feels like it could cause problems.

Reply:
It might be possible to lift the lookup and type printing into the TypeDescription constructor, but I wanted to avoid doing that because it makes TypeDescription a lot more complex in order to keep supporting the special handling for opaque result types and type aliases implemented in DiagnosticEngine, while still being trivially copyable.

I haven't been able to come up with a foolproof way to fix diagnostics type qualification that doesn't involve doing a lookup at some point. In retrospect, the `typeSpellingIsAmbiguous` check in DiagnosticEngine is only effective in a small number of cases.